resources/lib/playbackmanager.py

```diff
@@ -149,6 +149,8 @@ def extract_play_info(self, next_up_page, showing_next_up_page, showing_still_wa
             still_watching_page.close()
             should_play_default = still_watching_page.is_still_watching()
             should_play_non_default = still_watching_page.is_still_watching()
+        else:
+            return False, False
```

Comment:
I fail to understand when this would happen. I would expect this is run for either next-up or still-watching pop-ups, there should not be another scenario?

Reply:
It occurs if player is no longer running by the time show_popup_and_wait() is called in playbackmanager.py, in which case an exception is caught and show_popup_and_wait() returns without showing either of the pop-ups. This happens for example if the user skips forward past the end of the playing episode.

```
    def show_popup_and_wait(self, episode, next_up_page, still_watching_page):
        try:
            play_time = self.player.getTime()
            total_time = self.player.getTotalTime()
        except RuntimeError:
            self.log('exit early because player is no longer running', 2)
            return False, False
```

Reply:
As I said, this code doesn't read well, we are calling functions with information about both scenarios and then handling each individually in that function where it clearly should be two different cases throughout the code. That would make things more comprehensible.

But in this case I would not be expecting this function to be called when neither `showing_next_up_page` nor `showing_still_watching_page` is set.

Reply:
@thebertster Right, I was just going over the code and came to the same conclusion. And might this be related to #139 as well?

I was looking at:
```python
        while (self.player.isPlaying() and (total_time - play_time > 1)
               and not next_up_page.is_cancel() and not next_up_page.is_watch_now()
               and not still_watching_page.is_still_watching() and not still_watching_page.is_cancel()):
            try:
                play_time = self.player.getTime()
                total_time = self.player.getTotalTime()
            except RuntimeError:
                if showing_next_up_page:
                    next_up_page.close()
                    showing_next_up_page = False
                if showing_still_watching_page:
                    still_watching_page.close()
                    showing_still_watching_page = False
                break
```

Personally I think we need to make a different decision in this case. It would be much better if playbackmanager was more closely integrated with `Player()` so we can also handle `onPlayBackEnded` properly and handle it to play the next episode if it was configured this way. @im85288 you wrote this code, what do you think?

Reply:
@thebertster Maybe if you would add a note to that else-statement, I could agree with this workaround.
```suggestion
            # FIXME: This is a workaround until we handle this better (see comments in #142)
            return False, False
```

Reply:
I don't believe this is related to #139, as we can see the 'playing media episode' message being logged which is at the end of launch_popup() - we then just make the RPC call to play the media and I can't see any way that the playback itself would be aborted. I'm pretty sure this is a Kodi issue where a playback request occurs while the player is in the middle of shutting down...

Personally, I think PlaybackManager should be refactored so that the two pop-ups are handled as separate cases from the start, i.e. decide which of the two pop-ups needs to be displayed at the beginning of launch_popup() based on self.state.played_in_a_row and then branch to handle the two scenarios. There is still a lot of common code in show_popup_and_wait() that can be reused, but it seems ineligant to handle the two cases multiple times through the code.

I think it would make a lot of sense to let onPlaybackEnded trigger the RPC call to Kodi to play the next item. Then we can be sure the previous video player closed down fully. So instead of explicitly playing the next item, just always skip to the end of the current item, setting internal state as to what next item should be played when onPlaybackEnded is triggered (or if it is a playlist, Kodi should just play the next playlist item itself, right?).

However, if the event handling is properly multi-threaded, we could always have a potential race condition between us waiting for the end of the video and the onPlaybackEnded event firing if playback ends between us detecting the end of the video and setting the internal state as to what to play next. We'd need some careful synchronisation to prevent that in the worst case.

Reply:
Agreed in full. It's not easy, we have gone through this in the VRT NU add-on in order to get resumepoints updated in the cloud and it is messy. It is also something that is hard to build tests around to ensure it works in different scenarios (and e.g. different Kodi versions).